tools/cache.py

```diff
@@ -49,13 +50,15 @@ def acquire_cache_lock(self):
       self.prev_EM_EXCLUSIVE_CACHE_ACCESS = os.environ.get('EM_EXCLUSIVE_CACHE_ACCESS')
       os.environ['EM_EXCLUSIVE_CACHE_ACCESS'] = '1'
       logging.debug('Cache: done')
+    self.acquired_count += 1
 
   def release_cache_lock(self):
-    if not self.EM_EXCLUSIVE_CACHE_ACCESS:
+    if not self.EM_EXCLUSIVE_CACHE_ACCESS and self.acquired_count == 1:
```

Comment:
I think it would be more clear if we decrement the acquire count first, and then release if we get to 0, rather than releasing if the current count is 1.